### DISCUSSION TITLE
Merge PRs of multiple repositories

### DISCUSSION THREAD

**Discussion by eldadvcita** [Mergifyio/mergify#5051] [Q&A]
Hi,
My organization maintains multiple projects across multiple repositories.
Therefore, a feature release may require changes in more than one project / PRs.

As a developer, I would like to enqueue multiple PRs of different projects and have them merged only if all checks were successful.
Is such functionality/behavior currently supported?

**Comment by jd** [comment]
Hi @eldadvcita!

We currently have a `Depends-On` feature https://docs.mergify.com/actions/merge/#defining-pull-request-dependencies that allows specifying dependencies between PR (on the same repository). We have an item on our roadmap to extend this to multiple repositories, which would allow implementing of the workflow you describe in a linear way.

Do you think that's something that'd work for you?

**Reply by eldadvcita** [reply]
Hi @jd ,
Yes, I'm looking for a product that will support that exact feature (across different repositories).

**Reply by eldadvcita** [reply]
Since you plan to support such a feature, please take into consideration - integration checks.

- triggered once for all participating projects
- The trigger should include a list of projects, and a temp branch that contains all changes

In turn, CI process will be able to verify that all changes are working as expected.

**Reply by jd** [reply]
Could you elaborate a little bit about your problem in the first place? That'd make it easier to plan for the different pieces you might need. If you could describe your workflow that'd help.

In the meantime, I can talk about the way the feature is planned to work:
1. ProjectA has PR#A and ProjectB has PR#B
2. PR#B sets `Depends-On` on PR#A
3. If PR#B is ready to be merged, it won't be merged until PR#A is merged
4. Once PR#A is merged, if PR#B is still ready to be merged, it will be merged (or put in merge queue if merge queue is used)

Of course, that can work for a whole tree of PR dependencies.

**Reply by eldadvcita** [reply]
Hi @jd ,
Thank you for sharing how the feature will work. 
1. Is it possible to define more than one dependency?
2. Is it possible to interact with Mergify using PR comments?

> Could you elaborate a little bit about your problem in the first place
I would like to check the integrity of our ecosystem when introducing changes.
In other words, if a single PR or multiple PRs of multiple projects are about to be merged - I would like to conduct tests that will ensure that the new changes are working as expected. 

1. ProjectA has PR#A and ProjectB has PR#B & PR#C
2. Mergify creates predictive branches:
- ProjectA `branch_a` - `master` + PR#A
- ProjectB `branch_b` - `master` + PR#B + PR#C
3. Mergify trigger our CI, indicating which projects are about to change & corresponding predictive branches
4. CI setup a new environment. All projects are deployed from master except for ProjectA and ProjectB which are deployed using the predictive branch
5. CI conduct E2E, stress test, etc 
6. Mergify treats this flow like any other test


I hope it makes sense.

**Reply by jd** [reply]
@eldadvcita here are a few answers

1. Yes, this already works (in a single repo) with multiple dependencies, and the same will be true for external repositories when we get to it.
2. Yes, see [commands](https://docs.mergify.com/commands/).

> I would like to conduct tests that will ensure that the new changes are working as expected.

I think I understand but I'm not sure how Mergify would help with that. The way I see things, if you have a PR on ProjectA that depends on PR#B and PR#C you would do the following:

1. Run your CI job on PR#A
2. Detected that PR#A specifies 2 dependencies on ProjectB in your CI setup code
3. CI setup a new environment. All projects are deployed from `master` except for ProjectA and ProjectB which are deployed using the PR#A branch and when ProjectB is deployed you also `git pull` the PR#B and PR#C (because point 2.)
4. CI conduct E2E, stress test, etc and report success/failure on PR#A
5. Mergify treats this flow like any other test and decides if PR#A is mergeable or not (considering that anyway PR#B and PR#C needs to be merged first)

There is no need to setup any "predictive branches" from what I understand. The bulk of the work in the CI to test the indicated scenario in the PR ("I'm PR#A and I need PR#B and PR#C to work so please test with those included in the CI scenario").

Let me know if I miss something.
